<a href="https://colab.research.google.com/github/amandatz/linear-programming/blob/main/Atividade1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Atividade 1



Amanda Topanotti Zanette (22100776)

**Importações e funções auxiliares**

In [196]:
using LinearAlgebra, Printf

Esse trecho de código é usado apenas para garantir a repetibilidade dos números aleatórios gerados.

In [197]:
using Random
Random.seed!(42)

TaskLocalRNG()

## Questão 1

**Gradiente com backtracking**

In [198]:
function steepest_descent_backtracking(fun, x0; sigma=1e-4, rho=0.5, max_iter=2000, tol=1e-6)
  x = copy(x0)

  for k = 0:max_iter-1
    f, g, _ = fun(x)
    gnorm = norm(g)

    if gnorm <= tol
      return x, k, gnorm
    end

    # backtracking
    t = 1.0
    xn = x - t*g
    fn, _, _= fun(xn)

    while fn > f - sigma*t*gnorm^2
      t *= rho
      xn = x - t*g
      fn, _, _ = fun(xn)
    end

    x = xn
  end

  _, g, _ = fun(x)
  gnorm = norm(g)
  return x, max_iter, gnorm
end


steepest_descent_backtracking (generic function with 1 method)

**Newton com backtracking**

O método de Newton é dado por
$$
  x_{k+1} = x_k - \alpha_k H(x_k)^{-1} \nabla f(x_k)
$$
em que
- $\nabla f(x_k)$: gradiente de f;
- $H(x_k)$: matriz hessiana de f;
- $\alpha_k$: tamanho do passo.

In [199]:
function newton_backtracking(fun, x0; sigma=1e-4, rho=0.5, max_iter=2000, tol=1e-6)
  x = copy(x0)

  for k = 0:max_iter-1
    f, g, H = fun(x)
    gnorm = norm(g)

    if gnorm <= tol
      return x, k, gnorm
    end

    p = H \ g # VER ISSO DAQUI
    t = 1.0
    xn = x - t*p
    fn, _, _= fun(xn)

    # backtracking
    while fn > f - sigma*t*(g' * p)
      t *= rho
      xn = x - t*p
      fn, _, _ = fun(xn)
    end

    x = xn
  end

  f, g, H = fun(x)
  return x, max_iter, norm(g)
end


newton_backtracking (generic function with 1 method)

## Questão 2

Considere a função de Rosenbrock

In [200]:
function rosenb(x)
  a = 1.0
  b = 10.0

  f = (a - x[1])^2 + b*(x[2] - x[1]^2)^2

  grad_f = [ -2*(a - x[1]) + 2*b*(x[2] - x[1]^2)*(-2*x[1]),
        2*b*(x[2] - x[1]^2)]

  hess_f = [
        2 - 4*b*(x[2] - x[1]^2) + 8*b*x[1]^2   -4*b*x[1];
        -4*b*x[1]                               2*b
  ]

  return f, grad_f, hess_f
end

rosenb (generic function with 1 method)

Comparemos ambos os métodos anteriores considerando os pontos iniciais $x_0 = (1.01, 1.01)$ e $x_1 = (-1.0, 1.2)$

In [201]:
x0 = [1.01, 1.01]

x, max_iter_sd, gnorm_sd = steepest_descent_backtracking(rosenb, x0)
@printf("Gradiente:\n  Número de iterações: %d\n  Gradiente final: %.6e\n  Ponto final: [%.6f, %.6f]\n\n",
        max_iter_sd, gnorm_sd, x[1], x[2])

x, max_iter_n, gnorm_n = newton_backtracking(rosenb, x0)
@printf("Newton:\n  Número de iterações: %d\n  Gradiente final (norma): %.6e\n  Ponto final: [%.6f, %.6f]\n\n",
        max_iter_n, gnorm_n, x[1], x[2])


Gradiente:
  Número de iterações: 1027
  Gradiente final: 9.818612e-07
  Ponto final: [1.000001, 1.000002]

Newton:
  Número de iterações: 3
  Gradiente final (norma): 4.895830e-10
  Ponto final: [1.000000, 1.000000]



In [202]:
x1 = [-1.0, 1.2]

x, max_iter_sd, gnorm_sd = steepest_descent_backtracking(rosenb, x1)
@printf("Gradiente:\n  Número de iterações: %d\n  Gradiente final: %.6e\n  Ponto final: [%.6f, %.6f]\n\n",
        max_iter_sd, gnorm_sd, x[1], x[2])

x, max_iter_n, gnorm_n = newton_backtracking(rosenb, x1, max_iter=5000)
@printf("Newton:\n  Número de iterações: %d\n  Gradiente final (norma): %.6e\n  Ponto final: [%.6f, %.6f]\n\n",
        max_iter_n, gnorm_n, x[1], x[2]) # VER ISSO DAQUI

Gradiente:
  Número de iterações: 1184
  Gradiente final: 9.714477e-07
  Ponto final: [0.999999, 0.999998]

Newton:
  Número de iterações: 5000
  Gradiente final (norma): 5.656854e+00
  Ponto final: [-1.000000, 1.200000]



Para $x_0 = (1.01, 1.01)$ próximo do mínimo $x^* = (1,1)$, o método de Newton converge quadraticamente, resultando em apenas 3 iterações, enquanto o método do gradiente precisou de 1027 iterações. Além disso, o método de Newton foi bem mais preciso quando comparado ao método do gradiente.

Para $x_1 = (-1.0, 1.2)$, mais distante, o Newton falha em convergir (5000 iterações) porque $H(x)$ longe do mínimo pode não ser positiva definida, enquanto o gradiente ainda converge (1184 iterações), mas de forma linear, ilustrando que a convergência quadrática de Newton é local.

Observe que, para a função de Rosenbrock

$$
  f(x_1, x_2) = (x_1 - 1)^2 + 10 (x_2 - x_1^2)^2,
$$

a Hessiana é

$$
  H(x) =
  \begin{bmatrix}
  2 - 40(x_2 - x_1^2) + 80 x_1^2 & -40 x_1 \\
  -40 x_1 & 20
  \end{bmatrix}.
$$

e o determinante é

$$
  \det(H) = 40 - 800(x_2 - x_1^2).
$$

Para o ponto inicial $x_1 = (-1, 1.2)$:

$$
  x_2 - x_1^2 = 1.2 - 1 = 0.2 \implies \det(H) < 0,
$$

ou seja, a Hessiana **não é positiva definida** nesse ponto. Isso explica por que o método de Newton não converge a partir desse ponto inicial.


## Questão 3

In [203]:
function exact_line_search_quadratic(A, b, x, d)
  numerator = dot(b, d) - dot(d, A * x)
  denominator = dot(d, A * d)

  if denominator <= 0 || abs(denominator) < 1e-14
    return 1e-6   # passo mínimo seguro
  end
  return numerator / denominator
end

exact_line_search_quadratic (generic function with 1 method)

In [204]:
function quadratic(x, A, b)
  f = 0.5 * dot(x, A * x) - dot(b, x)
  g = A * x - b
  H = A
  return f, g, H
end

quadratic (generic function with 2 methods)

In [205]:
function householder_matrix(n)
  u = randn(n)
  u /= norm(u)
  U = Matrix(I, n, n) - 2.0 * (u * u')
  return U
end

householder_matrix (generic function with 1 method)

**Gradiente com busca linear exata**

In [206]:
function steepest_descent_exact(A, b, x0; max_iter=2000, tol=1e-6)
  x = copy(x0)

  for k = 0:max_iter-1
    f, g, _ = quadratic(x, A, b)

    gnorm = norm(g)
    if gnorm <= tol
      return x, k, gnorm
    end

    d = -g

    # Busca linear exata
    alpha = exact_line_search_quadratic(A, b, x, d)

    x = x + alpha * d
  end

  f, g, _ = quadratic(x, A, b)
  gnorm = norm(g)
  return x, max_iter, gnorm
end

steepest_descent_exact (generic function with 1 method)

**Newton com busca linear exata**

In [207]:
function newton_exact(A, b, x0; max_iter=2000, tol=1e-6)
  x = copy(x0)

  for k = 0:max_iter-1
    f, g, H = quadratic(x, A, b)

    gnorm = norm(g)
    if gnorm <= tol
      return x, k, gnorm
    end

    # direção de Newton
    p = - H \ g

    # passo exato
    alpha = exact_line_search_quadratic(A, b, x, p)

    x = x + alpha * p
  end

  f, g, H = quadratic(x, A, b)
  return x, max_iter, norm(g)
end

newton_exact (generic function with 2 methods)

Cria uma instância quadrática $f(x) = \tfrac{1}{2} x^T A x - b^T x$ de dimensão $n$ com uma distribuição específica de autovalores:  
- `:equal`: todos os autovalores iguais (bem condicionada).  
- `:two`: apenas dois valores diferentes (condicionamento moderado).  
- `:spread`: autovalores variados de 1 até 1000 (mal condicionada).

In [208]:
function generate_quadratic_instance(n, case)
  U = householder_matrix(n)

  eigenvalues = zeros(n)

  if case == :equal
    eigenvalues .= 1.0
  elseif case == :two
    eigenvalues[1:div(n,2)] .= 2.0
    eigenvalues[div(n,2)+1:end] .= 5.0
    shuffle!(eigenvalues)
  elseif case == :spread
    eigenvalues .= collect(LinRange(1.0, 1000.0, n))
    shuffle!(eigenvalues)
  else
    error("unkown case")
  end

  Lambda = Diagonal(eigenvalues)
  A = U * Lambda * U'
  A = 0.5 * (A + A')   # garante simetria

  b = randn(n)
  x0 = randn(n)

  return A, b, x0, eigenvalues
end

generate_quadratic_instance (generic function with 2 methods)

Neste bloco, aplicamos os métodos de otimização (`steepest_descent_exact` e `newton_exact`) em várias instâncias quadráticas geradas:

- `dimensions = [10, 100, 1000]`: diferentes tamanhos da matriz $A$.  
- `cases = [:equal, :two, :spread]`: diferentes distribuições de autovalores para testar o condicionamento.


In [210]:
dimensions = [10, 100, 1000]
cases = [:equal, :two, :spread]

for n in dimensions, case in cases
    A, b, x0, lambda = generate_quadratic_instance(n, case)

    # Gradiente
    t_sd = @elapsed x_sd, k_sd, gnorm_sd = steepest_descent_exact(A, b, x0, max_iter=10000)

    # Newton
    t_newt = @elapsed x_newt, k_newt, gnorm_newt = newton_exact(A, b, x0, max_iter=10000)

    println("Dimensão: $n, Caso: $case")
    println("  Gradiente:")
    println("    Iterações: $k_sd")
    println("    Tempo: $(round(t_sd, digits=4)) s")
    println("  Newton:")
    println("    Iterações: $k_newt")
    println("    Tempo: $(round(t_newt, digits=4)) s\n")
end

## ADICIONAR AVALIAÇÃO DE FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF

Dimensão: 10, Caso: equal
  Gradiente:
    Iterações: 1
    Tempo: 0.0 s
    Tempo: 0.0 s
  Newton:
    Iterações: 1
    Tempo: 0.0415 s

Dimensão: 10, Caso: two
  Gradiente:
    Iterações: 17
    Tempo: 0.0 s
    Tempo: 0.0 s
  Newton:
    Iterações: 1
    Tempo: 0.0 s

Dimensão: 10, Caso: spread
  Gradiente:
    Iterações: 6515
    Tempo: 0.0178 s
    Tempo: 0.0178 s
  Newton:
    Iterações: 1
    Tempo: 0.0 s

Dimensão: 100, Caso: equal
  Gradiente:
    Iterações: 1
    Tempo: 0.0 s
    Tempo: 0.0 s
  Newton:
    Iterações: 1
    Tempo: 0.0002 s

Dimensão: 100, Caso: two
  Gradiente:
    Iterações: 19
    Tempo: 0.0001 s
    Tempo: 0.0001 s
  Newton:
    Iterações: 1
    Tempo: 0.0001 s

Dimensão: 100, Caso: spread
  Gradiente:
    Iterações: 7012
    Tempo: 0.0569 s
    Tempo: 0.0569 s
  Newton:
    Iterações: 1
    Tempo: 0.0002 s

Dimensão: 1000, Caso: equal
  Gradiente:
    Iterações: 1
    Tempo: 0.0029 s
    Tempo: 0.0029 s
  Newton:
    Iterações: 1
    Tempo: 0.0374 s

Dimen

Observa-se que o método de Newton converge em apenas uma iteração para todos os casos testados, independentemente da dimensão da matriz ou da distribuição dos autovalores. Isso ocorre porque, para uma função quadrática, a direção de Newton aponta exatamente para o mínimo, e o passo exato leva diretamente a ele. O tempo de execução aumenta levemente com a dimensão devido à necessidade de resolver o sistema linear $H p = g$, mas permanece extremamente baixo.

Por outro lado, o desempenho do método do gradiente dependente do condicionamento da matriz $A$. Nos casos de autovalores iguais ou com apenas dois distintos, o número de iterações necessárias é relativamente baixo. Entretanto, quando os autovalores estão muito espalhados, o método precisa de mais iterações para convergir. O tempo de processamento cresce significativamente à medida que a dimensão da matriz aumenta.